app/src/main/java/io/github/zadam/triliumsender/LoginActivity.kt

```diff
@@ -42,129 +41,146 @@ class LoginActivity : AppCompatActivity() {
         if (TextUtils.isEmpty(username)) {
             usernameEditText.error = getString(R.string.error_field_required)
-            focusView = usernameEditText
-            cancel = true
+            usernameEditText.requestFocus()
+            return
         }
 
-        if (cancel) {
-            // There was an error; don't attempt login and focus the first
-            // form field with an error.
-            focusView?.requestFocus()
-        } else {
-            loginTask = UserLoginTask(triliumAddress, username, password)
-            loginTask!!.execute(null as Void?)
+        // Check for an empty password. Flag and abort if not.
+        if (TextUtils.isEmpty(password)) {
+            passwordEditText.error = getString(R.string.error_field_required)
+            passwordEditText.requestFocus()
+            return
         }
-    }
 
-    inner class LoginResult (val success: Boolean, val errorCode : Int?,
-                             val token : String? = null);
+        // Kick off a coroutine to handle the actual login attempt without blocking the UI.
+        // Since we want to be able to fire Toasts, we should use the Main (UI) scope.
+        val uiScope = CoroutineScope(Dispatchers.Main)
+        uiScope.launch {
 
-    /**
-     * Represents an asynchronous login/registration task used to authenticate
-     * the user.
-     */
-    inner class UserLoginTask internal constructor(private val triliumAddress: String, private val username: String, private val password: String) : AsyncTask<Void, Void, LoginResult>() {
+            val loginResult = doLogin(triliumAddress, username, password)
 
-        val TAG : String = "UserLoginTask"
+            if (loginResult.success) {
+                // Store the address and api token.
+                TriliumSettings(this@LoginActivity).save(triliumAddress, loginResult.token!!)
+                // Announce our success.
+                Toast.makeText(this@LoginActivity, "Trilium connection settings have been successfully configured.", Toast.LENGTH_LONG).show()
+                // End the activity.
+                finish()
+            } else {
+                if (loginResult.errorCode == R.string.error_network_error
+                        || loginResult.errorCode == R.string.error_unexpected_response) {
 
-        override fun doInBackground(vararg params: Void): LoginResult {
+                    triliumAddressEditText.error = getString(loginResult.errorCode)
+                    triliumAddressEditText.requestFocus()
+                } else if (loginResult.errorCode == R.string.error_incorrect_credentials) {
+                    passwordEditText.error = getString(loginResult.errorCode)
+                    passwordEditText.requestFocus()
+                } else {
+                    throw RuntimeException("Unknown code: " + loginResult.errorCode)
+                }
+            }
+        } // uiScope.launch {}
+    } // private fun AttemptLogin
```

Comment:
Just a habit to comment what context a large set of curly braces closes all at once-  this is probably the C++ dev in me coming through. Happy to delete these comments :)